File: Data Structures & Algorithms/count-subsequences/submission-2.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        # DP
        dp = {}

        def core(i, j):
            if j == len(t):
                return 1

            if i == len(s):
                return 0

            if (i, j) in dp:
                return dp[(i, j)]

            res = core(i + 1, j)
            if s[i] == t[j]:
                res += core(i + 1, j + 1)
            
            dp[(i, j)] = res
            return res

        return core(0, 0)

File: Data Structures & Algorithms/plus-one/submission-2.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int len = digits.size();
        int carry, total;
        vector<int> res;
        carry = total = 0;

        for (int i = len - 1; i >= 0; i--) {
            if (i == len - 1) {
                total = carry + digits[i] + 1;
            } else {
                total = carry + digits[i];
            }

            carry = total / 10;
            res.push_back(total % 10);
        }

        if (total >= 10) {
            res.push_back(total / 10);
        }

        reverse(res.begin(), res.end());
        return res;
    }
};
